### PR TITLE
libhns: Add support for XRC for HIP09

### PR DESCRIPTION
**Pull request by Li-Weihang**
The HIP09 supports XRC transport service, it greatly saves the number of
QPs required to connect all processes in a large cluster.

This series first adds support of ibv_create_qp/srq_ex which are necessary
interfaces for XRC, fixes an issue about CQ locks that is also required,
then XRC is fully supported at last.

The kernel part has been applied, you can refer it at: https://patchwork.kernel.org/project/linux-rdma/patch/1614826558-35423-1-git-send-email-liweihang@huawei.com/

### CODE REVIEW

File: providers/hns/hns_roce_u_verbs.c [hunk 1]
```diff
@@ -442,67 +483,124 @@ static int hns_roce_alloc_srq_buf(struct hns_roce_srq *srq)
...
+	if (is_basic_srq)
+		cmd_drv = &cmd.drv_payload;
+	else
+		cmd_drv = &cmd_ex.drv_payload;
+
+	cmd_drv->buf_addr = (uintptr_t)srq->buf.buf;
+	cmd_drv->que_addr = (uintptr_t)srq->idx_que.buf.buf;
+	cmd_drv->db_addr = (uintptr_t)srq->db;
+
+	if (is_basic_srq)
```
Comment (jgunthorpe) [inline providers/hns/hns_roce_u_verbs.c:559]: This if isn't needed always use cerate_srq_ex

Comment (Li-Weihang) [reply providers/hns/hns_roce_u_verbs.c]: Thank you, I will fix it

**Comment by rleon** [comment]
@jgunthorpe you requested changes, are we ok to merge it?

**Comment by Li-Weihang** [comment]
Hi Jason, is there anything that needs to be improved? @jgunthorpe

**Pull request merged**